File: bgm-code/ParameterEstimation_Stat/Convergence/Convergence_2.R
# clears workspace:
rm(list=ls(all=TRUE))

# sets working directories:
#setwd("F:/Codes/Convergence")
setwd("C:/Documents and Settings/ewagenmakers/My Documents/My Dropbox/EJ/temp/BayesBook/Code/ParameterEstimation_Stat/Convergence")
library(R2WinBUGS)
bugsdir = "C:/Program Files/WinBUGS14"

# Read data, a sequence of 33 participants * 90 problems 
# [3 probability estimates per problem] 
x = read.table("ConjunctionData.txt")

# Create three matrices: each row is a participant, each column is an item 
# Note that the [-30,] removes the data from participant 30: 
PA  = t(matrix(x[,1], nrow=90, ncol=33))[-30,]
PB  = t(matrix(x[,2], nrow=90, ncol=33))[-30,]
PAB = t(matrix(x[,3], nrow=90, ncol=33))[-30,]

# defines the number of items and number of participants
nsubj = length(PA[,1])
nitem = length(PA[1,])

data  = list("PA", "PB", "PAB", "nsubj", "nitem") # to be passed on to WinBUGS
myinits = list(
   list(betaphi = rnorm(nsubj,mean=0,sd=1), muphi = -0.6, sigmaphi = 0.2),
   list(betaphi = rnorm(nsubj,mean=0,sd=1), muphi = 0,   sigmaphi = 0.2),
   list(betaphi = rnorm(nsubj,mean=0,sd=1), muphi = 0.6,  sigmaphi = 0.2))

# parameters to be monitored:
parameters = c("beta", "muphi", "mu", "sigmaphi")

# The following command calls WinBUGS with specific options.
# For a detailed description see Sturtz, Ligges, & Gelman (2005).
samples = bugs(data, inits=myinits, parameters,
	 			model.file ="ConjunctionModel.txt",
	 			n.chains=3, n.iter=4000, n.burnin=1000, n.thin=1,
	 			DIC=T, bugs.directory=bugsdir,
	 			codaPkg=F, debug=T)
# Now the values for the monitored parameters are in the "samples" object, 
# ready for inspection.

# Assess Convergence with Coda package
library(coda)

coda.c1 = as.mcmc(samples$sims.array[,1,"mu"])
coda.c2 = as.mcmc(samples$sims.array[,2,"mu"])
coda.c3 = as.mcmc(samples$sims.array[,3,"mu"])

# Now put different chains in a single coda object
coda.c  = mcmc.list(coda.c1,coda.c2,coda.c3)

gelman.diag(coda.c)
# for more information, type "?gelman.diag" at the R prompt
# the help function also links to other stuff that coda can do
# Note that as a rule of thumb, Rhat must be < 1.1. The closer
# to 1.0 the better!
gelman.plot(coda.c)

# inspect the chains' autocorrelation (the lower the better):
autocorr(coda.c)

# the following allows you to easily "thin" the chain, leaving in
# only every thin'th value. This is not needed here, as the 
# autocorrelations are already very low. 
coda.c.thin = window(coda.c, thin=10)
autocorr(coda.c.thin)

# Visually check for fat hairy caterpillars and indistinguishability of the chains:
traceplot(coda.c)

File: bgm-code/ParameterEstimation_Stat/Convergence/Convergence_2.m
% Convergence 
clear;

% load data
load ConjunctionData;
x = ConjunctionData; 

%create data
%Reshape to matrix and transpose
PA = reshape(x(:,1),90,33)'; 
%Remove 30th participant
PA(30,:) = [];
PB = reshape(x(:,2),90,33)'; 
PB(30,:) = [];
PAB = reshape(x(:,3),90,33)'; 
PAB(30,:) = [];

% defines the number of items and number of participants
nsubj = length(PA(:,1));
nitem = length(PA(1,:));

% WinBUGS Parameters
nchains=3; % How Many Chains?
nburnin=1000; % How Many Burn-in Samples?
nsamples=4000; % How Many Recorded Samples?

% Assign Matlab Variables to the Observed WinBUGS Nodes
datastruct = struct('PA',PA,'PB',PB,'PAB',PAB,'nitem', nitem, 'nsubj',nsubj);

%vector of starting values for muphi
muphi = [-0.6,0,0.6];

% Initialize Unobserved Variables
for i=1:nchains
    S.betaphi = normrnd(0,1,1,nsubj);
    S.muphi = muphi(i);
    S.sigmaphi = 0.2;
    init0(i) = S;
end

% Use WinBUGS to Sample
[samples, stats] = matbugs(datastruct, ...
    fullfile(pwd, 'ConjunctionModel.txt'), ...
    'init', init0, ...
    'nChains', nchains, ...
    'view', 1, 'nburnin', nburnin, 'nsamples', nsamples, ...
    'thin', 1, 'DICstatus', 0, 'refreshrate',100, ...
    'monitorParams', {'beta','muphi','mu','sigmaphi'}, ...
    'Bugdir', 'C:/Program Files/WinBUGS14');




